pkg/apis/pipeline/v1/taskrun_validation_test.go

```diff
@@ -176,6 +244,36 @@ func TestTaskRun_Validate(t *testing.T) {
 			},
 		},
 		wc: config.EnableAlphaAPIFields,
+	}, {
+		name: "propagating partial object params in taskrun",
+		taskRun: &v1.TaskRun{
+			ObjectMeta: metav1.ObjectMeta{Name: "tr"},
+			Spec: v1.TaskRunSpec{
+				Params: []v1.Param{{
+					Name: "task-words",
+					Value: v1.ParamValue{
+						ObjectVal: map[string]string{"hello": "task run"},
+					},
+				}},
+				TaskSpec: &v1.TaskSpec{
+					Params: []v1.ParamSpec{{
+						Name: "task-words",
+						Type: v1.ParamTypeObject,
+						Default: &v1.ParamValue{
+							Type:      v1.ParamTypeObject,
+							ObjectVal: map[string]string{"hello": "task run def", "hi": "task"},
+						},
+					}},
+					Steps: []v1.Step{{
+						Name:    "echo",
+						Image:   "ubuntu",
+						Command: []string{"echo"},
+						Args:    []string{"$(params.task-words.hello)", "$(params.task-words.hi)"},
```

Comment:
this is interesting-- it seems like the task spec has defined a default value for the object param which contains the key "hi", and the default value is overridden by the value defined in the taskrun, which doesn't contain the key "hi". Why is it valid to use the key "hi" in the args? When a taskrun overrides the default value for an object param, is that on a per-key basis or for the param as a whole? Is this documented anywhere?

Reply:
That's a good question. Below is an example that should explain a few things. In the original [TEP](https://github.com/tektoncd/community/blob/main/teps/0107-propagating-parameters.md) on propagating params, we captured the usage for string and array types. But the main idea was that we wanted to have a [scoping mechanism](https://github.com/tektoncd/community/blob/main/teps/0107-propagating-parameters.md#scope-and-precedence). In this case, the keys provided at the task run will replace the default keys during substitution.  The scoping part is documented in the context of string and array types. Thanks for bringing this up. It seems like we should also update the documentation to show object replacements. 

```yaml
apiVersion: tekton.dev/v1beta1
kind: TaskRun
metadata:
...
spec:
  params:
  - name: gitrepo
    value:
      commit: sha123
      url: xyz.com
  serviceAccountName: default
  taskSpec:
    params:
    - default:
        branch: main
        commit: sha456
        url: abc.com
      name: gitrepo
      type: object
    steps:
    - args:
      - echo
      - --url=$(params.gitrepo.url)
      - --commit=$(params.gitrepo.commit)
      - --branch=$(params.gitrepo.branch)
...
status:
...
  taskSpec:
    params:
    - default:
        branch: main
        commit: sha456
        url: abc.com
      name: gitrepo
      type: object
    steps:
    - args:
      - echo
      - --url=xyz.com
      - --commit=sha123
      - --branch=main
...
```

Reply:
I think this is actually more related to TEP-0075 than TEP-0107. In TEP-0075, under [declaring object results and params](https://github.com/tektoncd/community/blob/main/teps/0075-object-param-and-result-types.md#declaring-object-results-and-params), it states "there will be no behavior to merge the default with the provided value, if for example the provided value specified some fields but not others". This sounds like if a user wants to replace the default value of an object param, any keys they do not specify will not have values, unlike the status.taskSpec.steps in your example.

@chuangw6 do you know if this behavior is validated anywhere? Would one of you be able to add documentation for it? I think it would also be reasonable to support the behavior shown in this example, but we would need to update TEP-0075 first.

Reply:
Yes, this is validated in [webhook](https://github.com/tektoncd/pipeline/blob/af4d0e6eb018a34c950180939a7f6a2f24941da6/pkg/apis/pipeline/v1beta1/task_validation.go#L408-L426). Since TEP75 proposes that `properties` section is required for either inline or ref definition of the spec, the validation makes sure the key used in steps is indeed declared in `properties`. The way I interpreted this line *"there will be no behavior to merge the default with the provided value, if for example the provided value specified some fields but not others"* in the TEP75 was that if an object `default` is provided, the value for all keys must be provided. Same for the value provided in taskrun. 

When it comes to propagating object params, I agree that it makes sense and provides more flexibility to propagate on   a per-key basis like the example @chitrangpatel shows above. @chitrangpatel also proposed that `properties` section is not required when it comes to propagating object param (like the example above), which I think makes sense too. This is because param propagating only happens for inline spec. As long keys being referenced exist in either `default` or `value`, it would be valid usage. (@chitrangpatel Please correct me if I'm wrong.)

So now it's about updating the documentation like TEP75 and [tasks.md](https://github.com/tektoncd/pipeline/blob/main/docs/tasks.md#parameter-type):
- the merge behaviour between `default` and `value` is allowed. And `value` from taskruns takes precedence. (FYI: #5270 might be related too.)
- `properties` section is not required when the spec is inline (embedded).

Did I miss anything here?

Reply:
Could you point me to any test cases that cover this?

I agree that having defaults overridden on a per-key basis is much more useful behavior, but that's not how I interpreted TEP-0075. It does look like #5270 is the feature we're describing. Can we update TEP-0075 as you've suggested?

No strong opinions about whether `properties` should be required in param propagation, but that should likely also be covered in TEP-0107.

Reply:
> Could you point me to any test cases that cover this?

Sure [here](https://github.com/tektoncd/pipeline/blob/af4d0e6eb018a34c950180939a7f6a2f24941da6/pkg/apis/pipeline/v1beta1/pipeline_validation_test.go#L1989-L2132) is the tests that cover this. I realized that the test for covering task steps was missed. I just added that https://github.com/tektoncd/pipeline/pull/5411. PTAL. Thanks @lbernick 


> I agree that having defaults overridden on a per-key basis is much more useful behavior, but that's not how I interpreted TEP-0075. It does look like https://github.com/tektoncd/pipeline/issues/5270 is the feature we're describing. Can we update TEP-0075 as you've suggested?

Sure. I'll update the TEP.

Reply:
@chitrangpatel I think this test case should result in an error until we resolve https://github.com/tektoncd/pipeline/pull/5427

Reply:
In that case, can we put this PR on hold until https://github.com/tektoncd/pipeline/pull/5427 is resolved? A lot of parameter propagation validation is around combining provided values with the defaults and comparing against the Properties field.

Reply:
sure, fine with me